### DISCUSSION TITLE
Performance overhead of using anyio

### DISCUSSION THREAD

**Discussion by Traktormaster** [agronholm/anyio#543] [General]
Hi!

### Preface
Firstly, I'd like to clarify that I've been developing asyncio applications and have not used trio/curio. I have not found the need to really consider using them so far, but different folks prefer one over the other for whatever reason, and that's fine. As far as I understand anyio can be used to make them work together in some capacity in the same program if two different dependencies use one or the other for example. For this reason some libraries have started to just migrate to anyio that have previously been asyncio-only.

### Problem
My problem with libraries depending on anyio is that the performance overhead of anyio can be significant while there is absolutely zero benefit for it, because only asncyio is ever utilized in my setups.

I've looked over the libraries that migrated over to anyio and anyio itself. I've quickly concluded that there are many great utilities that are available in anyio, so asking for an anyio-free version (usually by revert) is not even worth trying. After they've migrated to anyio, many custom utilities were dropped that would need to be re-introduced to the tree with tests, that most won't even consider, because anyio provides them. There is no way to convince said libraries to remove anyio for asyncio-only environments.

### Measurements
My main gripe is with the `get_asynclib` private function. It selects the "current" asynchronous backend that is in use, so it is necessarily used all the time. The following code focuses on the difference of using native asyncio code and this backend-selecing wrapper:

```python
import asyncio

import anyio


async def _busy_sleep_asyncio(cl):
    while True:
        await asyncio.sleep(0)
        cl[0] += 1


async def _busy_sleep_anyio(cl):
    while True:
        await anyio.sleep(0)
        cl[0] += 1


async def _check(fn, *args, duration=1):
    cl = [0]
    try:
        await asyncio.wait_for(asyncio.create_task(fn(*args, cl)), timeout=duration)
    except asyncio.TimeoutError:
        return cl[0]
    else:
        raise RuntimeError()


if __name__ == "__main__":
    direct = asyncio.run(_check(_busy_sleep_asyncio))
    print("asyncio:", direct)
    wrapped = asyncio.run(_check(_busy_sleep_anyio))
    print("anyio  :", wrapped)
    print("relative performance with anyio:", round(wrapped * 100 / direct, 2), "%")

```
The snippet above measures how many times the backend's sleep can be invoked with the native asyncio and the anyio-wrapped asyncio backend. On my machine the output is:
```
asyncio: 353484
anyio  : 250246
relative performance with anyio: 70.79 %
```
So compared to natively calling functions there can be an up to 29% performance penalty.

### Discussion
I'd like to initiate a discussion about how anyio selects the backend and how it could be improved.

1. Firstly I'd like to ask what a real use-case for selecting the backend dynamically looks like? I'd imagine there are two libraries in the same application and each use a different backend. After starting their own event loops, they can interact provided one or both utilize anyio?
2. If the selected backend could be frozen at the first call, the performance penalty would be more than halved. However, this may not be possible depending on what the answer for 1. might be.
3. I do not want to disturb anyone who uses anyio with my goals here necessarily. I'd also be interested in some global config override (by an environment variable maybe) that selects some alternate/optimized codes for asyncio-only behavior. It would be best that the selected backend is a global object that can be referenced directly, like the asyncio module itself.

**Comment by agronholm** [comment]
> Firstly I'd like to ask what a real use-case for selecting the backend dynamically looks like? I'd imagine there are two libraries in the same application and each use a different backend. After starting their own event loops, they can interact provided one or both utilize anyio?

The most common use case is a `trio` based app that needs to use a third party asyncio-only library. Thus, some tasks might use asyncio and some might be running `trio`.

I've been contemplating on providing a method of permanently selecting a back-end via an environment variable, at the obvious cost of compatibility when an app uses two back-ends, for whatever reason. On import, this would replace all the functions with their back-end specific versions, without going through `get_asynclib()`. Would this suffice?

My long term plan is to get rid of `trio` altogether by adopting its best parts in asyncio, but that will take a lot of effort and the will to push potentially breaking changes through to the standard library.

**Comment by smurfix** [comment]
> I've been contemplating on providing a method of permanently selecting a back-end via an environment variable

I'd prefer something like

```
import anyio
anyio.patch_for("trio")
```
(or `anyio.patch_for_trio()`), at the top of the main program / module / whatever. But aside from that this is a great idea.

> My long term plan is to get rid of trio altogether by adopting its best parts in asyncio

Good luck. You'll certainly need it.

**Comment by smurfix** [comment]
> Firstly I'd like to ask what a real use-case for selecting the backend dynamically looks like? I'd imagine there are two libraries in the same application and each use a different backend. After starting their own event loops, they can interact provided one or both utilize anyio?

If you want to use asyncio and trio in the same program, you need to use a Trio mainloop and supplement that with `trio_asyncio` and some judicious use of wrappers and/or compatibility hacks to switch from one mode to the other.

Method of last resort, frankly, as `trio_asyncio` is not exactly bug-free. It also pre-dates the ability to run Trio in "guest mode".

**Comment by agronholm** [comment]
> (or anyio.patch_for_trio()), at the top of the main program / module / whatever. But aside from that this is a great idea.

This becomes a problem if anything else has imported `anyio` first.

> Good luck. You'll certainly need it.

Don't I know that :smirk:

**Comment by smurfix** [comment]
> This becomes a problem if anything else has imported anyio first.

Well, that's why I wrote to put it at the top of the main program. There's been nobody yet to import anything from.

Besides, the environment variable needs to come from *somewhere*. You wouldn't want to require a separate wrapper just because your code uses an asyncio-only library and thus requires that backend; on the other hand, if your code sets that envvar before it imports anything (other than `os`…) you might as well go ahead and import+patch anyio there.

**Reply by davidbrochart** [reply]
> Well, that's why I wrote to put it at the top of the main program. There's been nobody yet to import anything from.

That's also how [nest-asyncio](https://github.com/erdewit/nest_asyncio) works:
```py
import nest_asyncio
nest_asyncio.apply()
```

**Comment by Traktormaster** [comment]
I'm glad to hear there is interest in this! The implementation specific selector function by an env-var would be the quickest and easiest improvement, but it does not eliminate the overhead completely. Freezing the backend in the selector improved the performance from 71% to 92%, this is somewhat similar to what would be expected of a contant-function selected by an env-var:
```python
_froze = None
def get_asynclib(asynclib_name: Optional[str] = None) -> Any:
    global _froze
    if _froze is None:
        if asynclib_name is None:
            asynclib_name = sniffio.current_async_library()

        modulename = "anyio._backends._" + asynclib_name
        try:
            _froze = sys.modules[modulename]
        except KeyError:
            _froze = import_module(modulename)

    return _froze
```

However, adapting my benchmark code a little can demonstrate that using the specific backend directly has zero overhead:
```python
import anyio._backends._asyncio as any_asyncio
async def _busy_sleep_anyio(cl):
    while True:
        await any_asyncio.sleep(0)
        cl[0] += 1
```

I would suggest that providing zero-overhead in general would be ideal. This could be possible if there was a new canonical import path for the backend like `asynio.backend` which could either be a direct import alias to a specific backend, or when the dynamic backend selection is desired could be a proxy-object which selects its `__dict__` to serve on attribute access.

**Reply by Traktormaster** [reply]
I've quickly put together a demo on how a dynamic proxy-module setup could be done: https://github.com/Traktormaster/dynmod-python

**Reply by agronholm** [reply]
> I'm glad to hear there is interest in this! The implementation specific selector function by an env-var would be the quickest and easiest improvement, but it does not eliminate the overhead completely.

I don't see why not. If this mechanism replaces imports in all the public modules directly with the functions from each backend, where does the extra overhead come from then?

**Reply by Traktormaster** [reply]
Compared to invoking the `asyncio.sleep` directly, the `anyio.sleep` is a wrapper sleep function that calls the backend selector that calls the sleep-implementation. If only the backend selector function is optimized, the extra two function calls still come with an ~5% overhead.

To have zero overhead the `anyio.sleep` has to actually be `asyncio.sleep` when the backend is selected as constant and not `anyio._core._eventloop.sleep` that calls a function to get the static backend to get the sleep implementation. I don't see a reason why this couldn't be done, the current sleep-wrapper function in anyio is just for the backend selection:
```python
async def sleep(delay: float) -> None:
    """
    Pause the current task for the specified duration.

    :param delay: the duration, in seconds

    """
    return await get_asynclib().sleep(delay)
```

The goal would be to invoke the sleep implementation on the backend selected either statically or dynamically since that is an alias for the real implementation. See the demo I've made: https://github.com/Traktormaster/dynmod-python

**Reply by agronholm** [reply]
What I was suggesting boils down to this:
```python
async def sleep(delay: float) -> None:
    """
    Pause the current task for the specified duration.

    :param delay: the duration, in seconds

    """
    return await asyncio.sleep(delay)
```
I don't imagine there is much overhead in this? Even this could be optimized to directly call `asyncio.sleep()`.

**Reply by Traktormaster** [reply]
Around 5% as I've stated:
```python
import asyncio

import anyio


async def _busy_sleep_asyncio(cl):
    while True:
        await asyncio.sleep(0)
        cl[0] += 1


async def _busy_sleep_anyio(cl):
    while True:
        await anyio.sleep(0)
        cl[0] += 1


async def sleep(delay: float) -> None:
    return await asyncio.sleep(delay)


async def _busy_sleep_wrapped(cl):
    while True:
        await sleep(0)
        cl[0] += 1


async def _check(fn, *args, duration=1):
    cl = [0]
    try:
        await asyncio.wait_for(asyncio.create_task(fn(*args, cl)), timeout=duration)
    except asyncio.TimeoutError:
        return cl[0]
    else:
        raise RuntimeError()


if __name__ == "__main__":
    direct = asyncio.run(_check(_busy_sleep_asyncio))
    print("asyncio:", direct)
    wrapped = asyncio.run(_check(_busy_sleep_anyio))
    print("anyio  :", wrapped)
    direct_wrap = asyncio.run(_check(_busy_sleep_wrapped))
    print("wrapped:", direct_wrap)
    print("relative performance with anyio:", round(wrapped * 100 / direct, 2), "%")
    print("relative performance with direct-wrapper:", round(direct_wrap * 100 / direct, 2), "%")

```

Output:
```
asyncio: 349538
anyio  : 248374
wrapped: 334808
relative performance with anyio: 71.06 %
relative performance with direct-wrapper: 95.79 %
```

**Reply by agronholm** [reply]
As a first step, 71.06 -> 95.79 % doesn't seem so bad, especially considering how little absolute overhead it has.

**Reply by Traktormaster** [reply]
Yes, it's a considerable improvement, that I'd be glad to have. A zero-overhead solution could be easier to create later, once this is done.

**Reply by agronholm** [reply]
Revisiting this, the script gives very inconsistent results. I found this script a bit more reliable:
```py
import asyncio
import time

import anyio


async def main(sleep_fn):
    start = time.monotonic_ns()
    for _ in range(100000):
        await sleep_fn(0)

    return time.monotonic_ns() - start


if __name__ == "__main__":
    asyncio_duration = asyncio.run(main(asyncio.sleep))
    print(f"asyncio: {asyncio_duration}")
    anyio_duration = asyncio.run(main(anyio.sleep))
    print(f"anyio: {anyio_duration}")
    print("anyio's performance overhead:", round((anyio_duration - asyncio_duration) / asyncio_duration * 100, 2), "%")
```
When run against the current `master`, I get a median overhead of ~45%.
When implementing "locking" the async backend, the overhead drops to ~17%.
If I skip `get_async_backend()` entirely and just call `asyncio.sleep()` directly from `anyio._eventloop.sleep()`, the overhead drops to ~6%.

I also ran just the anyio portion with `strace`. Without the async backend locking in place, I got 202574 system calls, while with locking, I got 102574 calls. So locking avoids roughly half of the system calls, at least in this somewhat contrived case.

I then started digging deeper and found that there were a lot of `getpid()` syscalls, triggered by `sniffio.current_async_library()`. It turns out that was due to `asyncio.get_running_loop()` needing to access thread locals.

**Reply by agronholm** [reply]
If you like, I have https://github.com/agronholm/anyio/pull/736 for you to evaluate. Let me know if you do try it.

**Comment by rafalkrupinski** [comment]
Would it be possible to make a light version of anyio, one that merely tries to import trio and falls back to asyncio, and just copies implementations straight from a backend?
Is there any downside to it?

**Reply by agronholm** [reply]
Yes, namely the inability to work at all with `trio-asyncio`.

**Reply by rafalkrupinski** [reply]
Obviously I don't know the code enough, but why code using anyio couldn't run with either backend when trio-asyncio is used?

**Reply by agronholm** [reply]
Because if you force it to use `trio`, then the asyncio code won't work, and vice versa. The same also happens if you try to run a test suite with asyncio and then trio (or vice versa), as the implementation is now locked to either one.

**Reply by rafalkrupinski** [reply]
Wait, that's not the case right now?

> Applications and libraries written against AnyIO’s API will run unmodified on **either** [asyncio](https://docs.python.org/3/library/asyncio.html) or [trio](https://github.com/python-trio/trio).

I understood I could run either asyncio + anyio code or trio + anyio...
it should really be O(1) cost to initialize it

**Reply by agronholm** [reply]
Every single one of my async libraries run the test suite on both, one at a time. But if I were to implicitly always choose either one and overwrite the functions, this would not work.

**Reply by agronholm** [reply]
Have you looked at #736? Won't it accomplish pretty much what you wanted?

**Reply by rafalkrupinski** [reply]
Then perhaps running tests could be improved

**Reply by rafalkrupinski** [reply]
> Have you looked at https://github.com/agronholm/anyio/pull/736? Won't it accomplish pretty much what you wanted?

Well, it's certainly positive that the overhead drops from 45 to 17%. Also I don't wish to sound ungrateful for this great code I'm getting for free :) We're just talking software engineering.

About dynamic selection of event loop - is it a project goal or just solving problem of tests?
> Applications and libraries written against AnyIO’s API will run unmodified on either [asyncio](https://docs.python.org/3/library/asyncio.html) or [trio](https://github.com/python-trio/trio).

> Every single one of my async libraries run the test suite on both, one at a time. But if I were to implicitly always choose either one and overwrite the functions, this would not work.

I don't mean to pick on what you've written here or elsewhere, it's just the impression I'm getting. Perhaps it's just a scope creep? IDK.

Personally I don't care much about this ability, I'd take the limitation and have near zero cost abstraction layer.
It's rather the dynamic backend selection that comes with 45% performance hit that should be opt-in (disregarding backwards compatibility for the sake of this conversation).
I'm not the only one, httpx is just migrating from anyio to what I wrote in my first comment - imports trio and falls back to asyncio. I'm afraid it will bite my a*s at some point. I'd rather have libraries use anyio and apps have the final say.

About #736: other than the above, it just feels wrong to have to communicate with a library via environment variables.

**Reply by agronholm** [reply]
Many people run `trio-asyncio`, even in production from what I've heard, so forcing one to pick either one at import time would be pissing on their use cases.

I have to ask you: what are you doing that sees such a significant overhead from AnyIO that even #736 does not fix to your satisfaction? Usually async apps spend most of their time waiting on I/O. The ~2 extra stack frames should not slow things down that much in practice.

**Reply by rafalkrupinski** [reply]
When using `trio-asyncio` anyio should run with either backend, no?

**Reply by agronholm** [reply]
The point of `trio-asyncio` is that it flip-flops between Trio and Asyncio, and in such cases, AnyIO needs to know the _current_ backend in order to choose the right calls to make.

**Reply by rafalkrupinski** [reply]
> The point of `trio-asyncio` is that it flip-flops between Trio and Asyncio, and in such cases, AnyIO needs to know the _current_ backend in order to choose the right calls to make.

Couldn't `trio-asyncio` be just another backend?

**Reply by agronholm** [reply]
> > The point of `trio-asyncio` is that it flip-flops between Trio and Asyncio, and in such cases, AnyIO needs to know the _current_ backend in order to choose the right calls to make.
> 
> Couldn't `trio-asyncio` be just another backend?

No, it doesn't work that way.

**Reply by rafalkrupinski** [reply]
Care to elaborate?

**Reply by agronholm** [reply]
Trio and asyncio both have their distinct APIs. AnyIO's backend selection determines which APIs it delegates the calls to from its own APIs. If the wrong API is chosen, it will fail, sometimes in unexpected ways.

**Reply by rafalkrupinski** [reply]
WDYT?
```python
# anyio/__init__.py

def install(impl: Literal['asyncio', 'trio']) -> None:
    if impl == 'asyncio':
        from ._backends._asyncio import AsyncIOBackend as Backend
    elif impl == 'trio':
        from ._backends._trio import TrioBackend as Backend
    else:
        raise ValueError(impl)

    target = sys.modules[__name__]

    for name in dir(Backend):
        if not name.startswith('_'):
            setattr(target, name, getattr(Backend, name))
```

it's still opt-in, just gets rid of `get_async_backend` completely.
I guess it's similar to other proposals...

**Reply by smurfix** [reply]
It's opt-in until it breaks.

Consider this setup (there are others).

Your program uses anyio, and you install the Trio backend. Your program also uses the anyio-supporting library A.

Library A needs to call B which is asyncio-only, so it uses sniffio to discover that it's on Trio, then wraps the calls to B with trio-asyncio.

B happens to have been ported to anyio in its latest release, no deprecations, so they didn't increase the major version number.

You update B.

Boom.

**Reply by smurfix** [reply]
Alternate case, no update required: in some circumstances (which you happen to not have tested for because frankly it's B's internals and far away from your app's code), B calls some code from library C, which uses anyio.

Boom.

**Reply by rafalkrupinski** [reply]
> It's opt-in until it breaks.

Lucky it's optional, when it breaks I can turn it off.

> Consider this setup (there are others).
> 
> Your program uses anyio, and you install the Trio backend. Your program also uses the anyio-supporting library A.
> 
> Library A needs to call B which is asyncio-only, so it uses sniffio to discover that it's on Trio, then wraps the calls to B with trio-asyncio.

OK, so my program uses trio and asyncio via a library... clearly I can't lock to either one, so I don't call `install()`

> B happens to have been ported to anyio in its latest release, no deprecations, so they didn't increase the major version number.
> 
> You update B.
>
> Boom.

Huh? Where did that come from?

**Reply by rafalkrupinski** [reply]
> Alternate case, no update required: in some circumstances (which you happen to not have tested for because frankly it's B's internals and far away from your app's code), B calls some code from library C, which uses anyio.

Let's see

- So my program runs in a trio loop, with anyio locked to `trio`
- uses library A that uses anyio, 
- A uses `trio-asyncio` to call asyncio-only B
- B calls C, which uses anyio, but B is in asyncio loop, while C, via `anyio`, is locked to trio

> Boom.

C'est la vie.

I promise I won't use trio in my apps ;)

**Reply by rafalkrupinski** [reply]
Perhaps it should not be called `install`, rather `lock_asyncio_and_blow_up_if_using_trio`, and vice versa

**Reply by agronholm** [reply]
> Perhaps it should not be called `install`, rather `lock_asyncio_and_blow_up_if_using_trio`, and vice versa

Perhaps _what_ should not be called `install`? And also, the mere _presence_ of `trio` should not affect how AnyIO initializes itself, as the user might want to use asyncio instead.

**Reply by rafalkrupinski** [reply]
> > Perhaps it should not be called `install`, rather `lock_asyncio_and_blow_up_if_using_trio`, and vice versa
> 
> Perhaps _what_ should not be called `install`? 

https://github.com/agronholm/anyio/discussions/543#discussioncomment-9981527

> And also, the mere _presence_ of `trio` should not affect how AnyIO initializes itself, as the user might want to use asyncio instead.

Naturally, locking to an implementation, whether via environment or function call should be done from `__main__` or equivalent.

**Reply by agronholm** [reply]
I don't understand what you're talking about here. Your original comment suggested trying to import trio, and then falling back to asyncio, and that's what I commented on.

**Reply by rafalkrupinski** [reply]
OK, let's try again...

What do you think of an alternative approach, a per [this code](https://github.com/agronholm/anyio/discussions/543#discussioncomment-9981527)

**Reply by agronholm** [reply]
> OK, let's try again...
> 
> What do you think of an alternative approach, a per [this code](https://github.com/agronholm/anyio/discussions/543#discussioncomment-9981527)

Well, of course it's not always as simple as replacing a free function with something from the async backend, but assuming it was, all we would gain is one less stack frame per call. Can you demonstrate that this makes enough of a real world difference that it's worth the effort?

**Reply by rafalkrupinski** [reply]
Oh, now I see how simple your change is.👍

**Reply by rafalkrupinski** [reply]
> It's opt-in until it breaks.
> 
> Consider this setup (there are others).
> 
> Your program uses anyio, and you install the Trio backend. Your program also uses the anyio-supporting library A.
> 
> Library A needs to call B which is asyncio-only, so it uses sniffio to discover that it's on Trio, then wraps the calls to B with trio-asyncio.
> 
> B happens to have been ported to anyio in its latest release, no deprecations, so they didn't increase the major version number.
> 
> You update B.
> 
> Boom.

@smurfix 
These are interesting problems you've brought up

The first issue is an obvious bug in A - an anyio library that might be running under trio is using asyncio code. To me it's the same kind of a problem as a library trying to read a non-existing file and raising unexpected FileNotFoundError.

The second issue is a problem with B - a significant change in behaviour without increasing the major number.

I think it's alright to protect users from such bugs _if_ it's either cheap or it's opt-in.

**Comment by smurfix** [comment]
On 07.07.24 23:08, Raphael Krupinski wrote:
> When using |trio-asyncio| anyio should run with either backend, no?

It doesn't know whether there's a Trio runner in control of the asyncio 
"loop", and testing for that situation would add even more overhead.

-- 
-- mit freundlichen Grüßen
-- 
-- Matthias Urlichs

**Reply by rafalkrupinski** [reply]
> On 07.07.24 23:08, Raphael Krupinski wrote: When using |trio-asyncio| anyio should run with either backend, no?
> It doesn't know whether there's a Trio runner in control of the asyncio "loop", and testing for that situation would add even more overhead.

I made that comment before I knew anything about trio-asyncio, based only on the header in its documentation:
`trio-asyncio: A re-implementation of the asyncio mainloop on top of Trio`

Of course, if you actually have two event loops and bridge them with `trio-asyncio`, locking wouldn't work

**Comment by jruere** [comment]
Hi! I'm new to this discussion, so apologies in advance for my likely confusion.

I understand that while using trio, asyncio could also be indirectly in use, so we cannot avoid the per-call backend dispatch.

Is this also possible when using asyncio? Could trio be also in use?

**Reply by smurfix** [reply]
In a different thread, or with Trio guest mode, yes.

AFAIK nobody has written a trio-on-asyncio guest mode module yet, but there's no reason why that can't be done.

**Reply by jruere** [reply]
The price of being a super dynamic language, I guess.

How about expecting a single backend but allowing the user to configure AnyIO to remain dynamic? This could be sensible if 90% of the users would benefit and the users of these mixed backends would not be unexpected (e.g. a module just does this, and you never knew).

Alternatively, caching results based on something cheap to calculate. Is there a Task ID? Something like that?
If a Task were always run in the event loop where it was created, then perhaps the Task could have a reference to its Backend?

**Reply by agronholm** [reply]
In my own tests, only pathological, unrealistic benchmarks show any notable performance improvements when `sniffio` is skipped, so I'm not 100% sure this is something we want to focus on. AnyIO does have performance issues elsewhere. I just tested the performance of `anyio.Lock` and it's much slower than the respective native implementations.

**Reply by jruere** [reply]
The PRs referenced from https://github.com/agronholm/anyio/discussions/760 have a benchmark of HTTPX, which is greatly affected by synchronization primitives.

**Reply by agronholm** [reply]
Here I have a PR that pretty much eliminates the locking overhead of AnyIO vs asyncio: https://github.com/agronholm/anyio/pull/761
